File: src/main/java/org/lanternpowered/server/service/permission/DataFactoryCollection.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.service.permission;

import static com.google.common.base.Preconditions.checkNotNull;

import org.lanternpowered.server.service.permission.base.LanternSubject;
import org.lanternpowered.server.service.permission.base.LanternSubjectCollection;
import org.spongepowered.api.service.context.Context;
import org.spongepowered.api.service.permission.MemorySubjectData;
import org.spongepowered.api.service.permission.PermissionService;
import org.spongepowered.api.service.permission.Subject;
import org.spongepowered.api.service.permission.SubjectCollection;
import org.spongepowered.api.service.permission.SubjectReference;
import org.spongepowered.api.util.Tristate;

import java.util.Collection;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.function.Function;

final class DataFactoryCollection extends LanternSubjectCollection {

    private final LanternPermissionService service;
    private final ConcurrentMap<String, LanternSubject> subjects = new ConcurrentHashMap<>();
    private final Function<String, MemorySubjectData> dataFactory;

    DataFactoryCollection(String identifier, LanternPermissionService service,
            Function<String, MemorySubjectData> dataFactory) {
        super(identifier, service);
        this.dataFactory = dataFactory;
        this.service = service;
    }

    @Override
    public LanternSubject get(String identifier) {
        checkNotNull(identifier, "identifier");
        return this.subjects.computeIfAbsent(identifier, id -> new DataFactorySubject(id, this.dataFactory.apply(id)));
    }

    @Override
    public boolean isRegistered(String identifier) {
        checkNotNull(identifier, "identifier");
        return this.subjects.containsKey(identifier);
    }

    @Override
    @SuppressWarnings("unchecked")
    public Collection<Subject> getLoadedSubjects() {
        return (Collection) this.subjects.values();
    }

    private final class DataFactorySubject extends LanternSubject {

        private final String identifier;
        private final MemorySubjectData data;

        DataFactorySubject(String identifier, MemorySubjectData data) {
            this.identifier = identifier;
            this.data = data;
        }

        @Override
        public String getIdentifier() {
            return this.identifier;
        }

        @Override
        public Optional<String> getFriendlyIdentifier() {
            return Optional.empty();
        }

        @Override
        public SubjectCollection getContainingCollection() {
            return DataFactoryCollection.this;
        }

        @Override
        public PermissionService getService() {
            return DataFactoryCollection.this.service;
        }

        @Override
        public SubjectReference asSubjectReference() {
            return DataFactoryCollection.this.service.newSubjectReference(DataFactoryCollection.this.getIdentifier(), getIdentifier());
        }

        @Override
        public MemorySubjectData getSubjectData() {
            return this.data;
        }

        @Override
        public Tristate getPermissionValue(Set<Context> contexts, String permission) {
            Tristate ret = super.getPermissionValue(contexts, permission);
            if (ret == Tristate.UNDEFINED) {
                ret = getDataPermissionValue(DataFactoryCollection.this.getDefaults().getTransientSubjectData(), permission);
            }
            if (ret == Tristate.UNDEFINED) {
                ret = getDataPermissionValue(DataFactoryCollection.this.service.getDefaults().getTransientSubjectData(), permission);
            }
            return ret;
        }

        @Override
        public Optional<String> getOption(Set<Context> contexts, String option) {
            Optional<String> ret = super.getOption(contexts, option);
            if (!ret.isPresent()) {
                ret = getDataOptionValue(DataFactoryCollection.this.getDefaults().getSubjectData(), option);
            }
            if (!ret.isPresent()) {
                ret = getDataOptionValue(DataFactoryCollection.this.service.getDefaults().getSubjectData(), option);
            }
            return ret;
        }
    }
}

File: src/main/kotlin/org/lanternpowered/server/service/permission/DataFactoryCollection.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.service.permission

import org.lanternpowered.api.util.collections.concurrentHashMapOf
import org.lanternpowered.api.util.optional.emptyOptional
import org.spongepowered.api.service.context.Context
import org.spongepowered.api.service.permission.MemorySubjectData
import org.spongepowered.api.service.permission.PermissionService
import org.spongepowered.api.service.permission.Subject
import org.spongepowered.api.service.permission.SubjectCollection
import org.spongepowered.api.service.permission.SubjectReference
import org.spongepowered.api.util.Tristate
import java.util.Optional

internal class DataFactoryCollection(
        identifier: String,
        private val service: LanternPermissionService,
        private val dataFactory: (String) -> MemorySubjectData
) : LanternSubjectCollection(identifier, service) {

    private val subjects = concurrentHashMapOf<String, LanternSubject>()

    override fun get(identifier: String): LanternSubject =
            this.subjects.computeIfAbsent(identifier) { id -> DataFactorySubject(id, this.dataFactory(id)) }

    override fun isRegistered(identifier: String): Boolean = this.subjects.containsKey(identifier)
    override fun getLoadedSubjects(): Collection<Subject> = this.subjects.values

    private inner class DataFactorySubject internal constructor(
            private val identifier: String,
            private val data: MemorySubjectData
    ) : LanternSubject() {

        override val service: PermissionService
            get() = this@DataFactoryCollection.service

        override fun getIdentifier(): String = this.identifier
        override fun getFriendlyIdentifier(): Optional<String> = emptyOptional()
        override fun getContainingCollection(): SubjectCollection = this@DataFactoryCollection
        override fun getSubjectData(): MemorySubjectData = this.data

        override fun asSubjectReference(): SubjectReference =
                this@DataFactoryCollection.service.newSubjectReference(this@DataFactoryCollection.identifier, this.getIdentifier())

        override fun getPermissionValue(contexts: Set<Context>, permission: String): Tristate {
            var result = super.getPermissionValue(contexts, permission)
            if (result == Tristate.UNDEFINED)
                result = this.getDataPermissionValue(this@DataFactoryCollection.defaults.transientSubjectData, permission)
            if (result == Tristate.UNDEFINED)
                result = this.getDataPermissionValue(this@DataFactoryCollection.service.defaults.transientSubjectData, permission)
            return result
        }

        override fun getOption(contexts: Set<Context>, key: String): Optional<String> {
            var result = super.getOption(contexts, key)
            if (!result.isPresent)
                result = this.getDataOptionValue(this@DataFactoryCollection.defaults.subjectData, key)
            if (!result.isPresent)
                result = this.getDataOptionValue(this@DataFactoryCollection.service.defaults.subjectData, key)
            return result
        }
    }
}
